syntax = "proto3";

package todos.quickstart;

import "google/protobuf/empty.proto";

service PostController {
    rpc Create(PostRequest) returns (PostResponse) {}
    rpc Destroy(PostDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(PostListRequest) returns (PostListResponse) {}
    rpc PartialUpdate(PostPartialUpdateRequest) returns (PostResponse) {}
    rpc Retrieve(PostRetrieveRequest) returns (PostResponse) {}
    rpc Update(PostRequest) returns (PostResponse) {}
}

service UserActionController {
    rpc Create(UserRequest) returns (UserResponse) {}
    rpc Destroy(UserDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(UserListRequest) returns (UserListResponse) {}
    rpc PartialUpdate(UserPartialUpdateRequest) returns (UserResponse) {}
    rpc Retrieve(UserRetrieveRequest) returns (UserResponse) {}
    rpc Update(UserRequest) returns (UserResponse) {}
}

service UserController {
    rpc List(UserListRequest) returns (UserListResponse) {}
    rpc Retrieve(UserRetrieveRequest) returns (UserResponse) {}
}

message PostDestroyRequest {
    int32 id = 1;
}

message PostListRequest {
}

message PostListResponse {
    repeated PostResponse results = 1;
}

message PostPartialUpdateRequest {
    optional int32 id = 1;
    repeated string _partial_update_fields = 2;
    string headline = 3;
    string content = 4;
    int64 user = 5;
}

message PostRequest {
    optional int32 id = 1;
    string headline = 2;
    string content = 3;
    int64 user = 4;
}

message PostResponse {
    optional int32 id = 1;
    optional string pub_date = 2;
    string headline = 3;
    string content = 4;
    int64 user = 5;
}

message PostRetrieveRequest {
    int32 id = 1;
}

message UserDestroyRequest {
    int32 id = 1;
}

message UserListRequest {
}

message UserListResponse {
    repeated UserResponse results = 1;
}

message UserPartialUpdateRequest {
    optional int32 id = 1;
    string full_name = 2;
    repeated string _partial_update_fields = 3;
}

message UserRequest {
    optional int32 id = 1;
    string full_name = 2;
}

message UserResponse {
    optional int32 id = 1;
    string full_name = 2;
}

message UserRetrieveRequest {
    int32 id = 1;
}

